### PR TITLE
Integrating React Point Cloud Viewer

### PR DESCRIPTION
**Pull request by adisingh50**
**Note:** please **ignore this PR**. I will be piece mealing it into 3 smaller PRs and having those approved and merged into master sequentially. I will delete this branch once all the React code has been checked into master via the 3 smaller PRs.

### CODE REVIEW

**Comment by johnwlambert** [comment]
Thanks for the PR, @adisingh50 . I'll take a look!

File: gtsfm/common/view_frustum.py [hunk 1]
```diff
@@ -57,7 +57,8 @@ def get_frustum_vertices_camfr(
             ]
         )
 
-        ray_dirs = compute_pixel_ray_directions_vectorized(uv, self.fx_, self.img_w_, self.img_h_)
+        ray_dirs = compute_pixel_ray_directions_vectorized(
```
Comment (johnwlambert) [inline gtsfm/common/view_frustum.py:60]: i think this change is unrelated -- do you mind checking out the change here/resetting this file to HEAD (carefully :-) though)

File: gtsfm/multi_view_optimizer.py [hunk 2]
```diff
@@ -24,6 +24,7 @@
 from gtsfm.bundle.bundle_adjustment import BundleAdjustmentOptimizer
 from gtsfm.data_association.data_assoc import DataAssociation
 
+RESULT_METRICS_REACT_PATH = Path(__file__).resolve().parent.parent / "rtf_vis_tool" / "src" / "result_metrics"
```
Comment (johnwlambert) [inline gtsfm/multi_view_optimizer.py:27]: maybe an inline comment here above would help explain why we need to save in 2 places?

File: rtf_vis_tool/.eslintcache [hunk 3]
```diff
@@ -0,0 +1 @@
...Column":34},"no-unused-vars","'Canvas' is defined but never used.","Identifier","unusedVar","'PCViewer' is defined but never used.","'TranSuccessSummary' is defined but never used.","'showTSS' is assigned a value but never used.","'toggleTranSummaryDisplay' is assigned a value but never used.","react/jsx-pascal-case","Imported JSX component Data_Association_PC must be in PascalCase or SCREAMING_SNAKE_CASE","JSXOpeningElement","'useEffect' is defined but never used.","'useRef' is defined but never used.","eqeqeq","Expected '===' and instead saw '=='.","BinaryExpression","unexpected","no-throw-literal","Expected an error object to be thrown.","ThrowStatement","object","'SE3' is assigned a value but never used.","'SpriteMesh' is defined but never used.","react-hooks/exhaustive-deps","React Hook useEffect has a missing dependency: 'props.json'. Either include it or remove the dependency array.","ArrayExpression",["145"],"array-callback-return","Array.prototype.map() expects a return value from arrow function.","ArrowFunctionExpression","expectedInside",{"desc":"146","fix":"147"},"Update the dependencies array to be: [props.json]",{"range":"148","text":"149"},[1598,1600],"[props.json]"]
```
Comment (johnwlambert) [inline rtf_vis_tool/.eslintcache:1]: these look like files specific to your system Adi -- do we need them?
```python
"C:\\Users\\adisi\\Desktop\\Georgia Tech\\Spring 2021\\SfM Research\\gtsfm\\rtf_vis_tool\\src\\Components\\SfMDataDivNode.js"
```

File: rtf_vis_tool/.gitignore [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
+
+# dependencies
```
Comment (johnwlambert) [inline rtf_vis_tool/.gitignore:3]: can we share this with the .gitignore in the top-level directory?

would eslintcache go in here?

File: rtf_vis_tool/public/index.html [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8" />
+    <title>GTSFM Pipeline Visualizer</title>
+  </head>
+  <body>
+		<!-- <script src="lib/threejs/OrbitControls.js"></script> -->
```
Comment (johnwlambert) [inline rtf_vis_tool/public/index.html:8]: is this commented out code? can we remove it?

File: rtf_vis_tool/package-lock.json [hunk 6]
```diff
@@ -0,0 +1,17982 @@
+{
+  "name": "rtf-intro",
+  "version": "0.1.0",
+  "lockfileVersion": 1,
+  "requires": true,
+  "dependencies": {
```
Comment (johnwlambert) [inline rtf_vis_tool/package-lock.json:6]: there are a lot of dependencies :-) 

Do we need every one of them? What are they fore?

File: rtf_vis_tool/src/result_metrics/multiview_optimizer_metrics.json [hunk 7]
```diff
@@ -0,0 +1,113 @@
+{
+    "rotation_averaging_angle_deg": {
+        "median_error": 2.915834376502234,
```
Comment (johnwlambert) [inline rtf_vis_tool/src/result_metrics/multiview_optimizer_metrics.json:3]: this looks like sample data, that probably belongs in a unit test. Is there a way to use them in a unit test?

File: rtf_vis_tool/yarn.lock [hunk 8]
```diff
@@ -0,0 +1,12315 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
+# yarn lockfile v1
+
+
+"@alloc/types@^1.2.1":
```
Comment (johnwlambert) [inline rtf_vis_tool/yarn.lock]: this looks auto-generated. do we need to check it in, or is created when the code is run?

File: rtf_vis_tool/src/Components/PCViewer.js [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+            v2={[verts_worldfr[2].get(0,0),verts_worldfr[2].get(0,1),verts_worldfr[2].get(0,2)]} 
+            v3={[verts_worldfr[3].get(0,0),verts_worldfr[3].get(0,1),verts_worldfr[3].get(0,2)]} 
+            v4={[verts_worldfr[4].get(0,0),verts_worldfr[4].get(0,1),verts_worldfr[4].get(0,2)]}
+            width={0.5}/>);
+        }
+        setFrustums(finalFrustumsJSX);
+    }
+    
+    //Function
+    const visualize_3D_Points_File = () => {
```
Comment (johnwlambert) [inline rtf_vis_tool/src/Components/PCViewer.js:121]: missing docstring here

File: rtf_vis_tool/src/Components/PCViewer.js [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+        const reader = new FileReader();
+        reader.readAsText(e.target.files[0]);
+        reader.onload = (e) => {
+        const fileString = e.target.result;
+        setRawExtrinsicString(fileString);
+        }
+    }
+
+    //Function
+    const visualize_camera_frustums = () => {
```
Comment (johnwlambert) [inline rtf_vis_tool/src/Components/PCViewer.js:60]: missing docstring here

File: rtf_vis_tool/src/Components/PCViewer.js [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+        const reader = new FileReader();
+        reader.readAsText(e.target.files[0])
+        reader.onload = (e) => {
+        const fileString = e.target.result;
+        setRawFileString(fileString);
+        }
+    }
+
+    //Function
+    const readIntrinsicsFile = (e) => {
```
Comment (johnwlambert) [inline rtf_vis_tool/src/Components/PCViewer.js:38]: docstring

File: rtf_vis_tool/src/Components/PCViewer.js [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        e.preventDefault();
+        const reader = new FileReader();
+        reader.readAsText(e.target.files[0]);
+        reader.onload = (e) => {
+        const fileString = e.target.result;
+        setRawIntrinsicString(fileString);
+        }
+    }
+
+    //Function
```
Comment (johnwlambert) [inline rtf_vis_tool/src/Components/PCViewer.js:48]: docstring

File: rtf_vis_tool/src/Components/PCViewer.js [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+const PCViewer = (props) => {
+    const [rawFileString, setRawFileString] = useState("");
+    const [rawIntrinsicString, setRawIntrinsicString] = useState("");
+    const [rawExtrinsicString, setRawExtrinsicString] = useState("");
+    const [pointCloud, setPointCloud] = useState([]);
+    const [frustums, setFrustums] = useState([]);
+    const [numCams, setNumCams] = useState(0);
+
+    //Function
```
Comment (johnwlambert) [inline rtf_vis_tool/src/Components/PCViewer.js:26]: docstring

File: rtf_vis_tool/src/Components/DivGraph.js [hunk 14]
```diff
@@ -0,0 +1,146 @@
+import React, {useEffect, useState} from "react";
+import Xarrow from "react-xarrows";
+
+import EdgeList from './gtsfm_edge_list.js';
```
[the remark this answers is not in the document]
Comment (johnwlambert) [reply rtf_vis_tool/src/Components/DivGraph.js:4]: (within groups)

File: rtf_vis_tool/src/Components/Data_Association_PC.js [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+import SpriteMesh from './SpriteMesh';
+import CoordinateGrid from './CoordinateGrid';
+import OrbitControlsComponent from './OrbitControlsComponent';
+import PointMesh from './PointMesh';
+
+
+extend({OrbitControls})
+
+const Data_Association_PC = (props) => {
```
Comment (johnwlambert) [inline rtf_vis_tool/src/Components/Data_Association_PC.js:14]: docstring needed

File: gtsfm/common/view_frustum.py (hunk 1 above)
Comment (adisingh50) [reply gtsfm/common/view_frustum.py:60]: oops. I think my VScode auto formatter did this. Will reset this back to original.

File: package-lock.json [hunk 16]
```diff
@@ -0,0 +1,11 @@
+{
```
[the remark this answers is not in the document]
Comment (adisingh50) [reply package-lock.json:1]: Actually, I can remove this altogether.

File: rtf_vis_tool/.eslintcache (hunk 3 above)
Comment (adisingh50) [reply rtf_vis_tool/.eslintcache:1]: Nope. I'll add this to my .gitignore

File: rtf_vis_tool/package-lock.json (hunk 6 above)
Comment (adisingh50) [reply rtf_vis_tool/package-lock.json:6]: to be honest, a lot of the stuff in package-lock.json comes auto generated when I initialize my react application. I'll look into if and how I can safely clean this file up. Because you're right, it's extremely long...

File: rtf_vis_tool/yarn.lock (hunk 8 above)
Comment (adisingh50) [reply rtf_vis_tool/yarn.lock]: after reading this [forum](https://stackoverflow.com/questions/44552348/should-i-commit-yarn-lock-and-package-lock-json-files) online, I believe both yarn.lock and package-lock.json files will need to be committed to ensure all users have the same Node dependencies.

File: rtf_vis_tool/package-lock.json (hunk 6 above)
Comment (adisingh50) [reply rtf_vis_tool/package-lock.json:6]: Was able to clean some of it up, but this package-lock.json file will have to be checked into Github to ensure all users are using the same package versions.

File: rtf_vis_tool/src/result_metrics/multiview_optimizer_metrics.json (hunk 7 above)
Comment (adisingh50) [reply rtf_vis_tool/src/result_metrics/multiview_optimizer_metrics.json:3]: will remove all the files in result_metrics since those are generated anyways after running scene_optimizer.py. Will look more into unit testing.

File: gtsfm/multi_view_optimizer.py [hunk 17]
```diff
@@ -24,6 +24,8 @@
 from gtsfm.bundle.bundle_adjustment import BundleAdjustmentOptimizer
 from gtsfm.data_association.data_assoc import DataAssociation
 
+#used to save result_metrics files within React application for internal rendering
+RESULT_METRICS_REACT_PATH = Path(__file__).resolve().parent.parent / "rtf_vis_tool" / "src" / "result_metrics"
```
Comment (ayushbaid) [inline gtsfm/multi_view_optimizer.py:28]: I may have missed it, but are we creating two copies of the files? Or are we creating symlinks?

**Comment by johnwlambert** [comment]
@adisingh50, do you mind merging in the latest copy of master into your branch?

File: gtsfm/multi_view_optimizer.py (hunk 17 above)
Comment (adisingh50) [reply gtsfm/multi_view_optimizer.py:28]: So the problem with directly importing the result_metric files in the gtsfm folder into my React application is that React will not let me import anything outside it's source code folder. Because of this limitation, the next best option would be to save the result_metrics files in twice. Once in the gtsfm folder like normal and once within my React project folder.

**Comment by johnwlambert** [comment]
@adisingh50 try to find a way to split the PR into 3 chunks.

**Comment by ayushbaid** [comment]
Is this PR outdated now?

**Comment by adisingh50** [comment]
yes it's outdated. This PR will definitely not be merged into master. I will close this PR now , but still keep the `reactVisTool` branch alive until all the React code is merged into master via other smaller PRs.
